iothub/device/src/Edge/EdgeModuleClientHelper.cs

```diff
@@ -103,27 +71,65 @@ private async Task<IotHubModuleClient> CreateInternalClientFromEnvironmentAsync(
+        {
+            // Make sure client options is initialized.
+            if (options == default)
+            {
+                options = new();
+            }
+
+            ICertificateValidator certificateValidator = NullCertificateValidator.Instance;
+
+            IDictionary envVariables = Environment.GetEnvironmentVariables();
+
+            string connectionString = GetValueFromEnvironment(envVariables, EdgehubConnectionstringVariableName)
+                ?? GetValueFromEnvironment(envVariables, IothubConnectionstringVariableName);
+
+            // First try to create from connection string and if env variable for connection string is not found try to create from edgedUri
+            if (!string.IsNullOrWhiteSpace(connectionString))
+            {
+                string certPath = Environment.GetEnvironmentVariable(EdgeCaCertificateFileVariableName);
+
+                if (!string.IsNullOrWhiteSpace(certPath))
+                {
+                    Debug.WriteLine("EdgeModuleClientFactory setupTrustBundle from file");
+                    var expectedRoot = new X509Certificate2(certPath);
+                    certificateValidator = GetCertificateValidator(new List<X509Certificate2>() { expectedRoot }, options);
+                }
+
+                return certificateValidator;
+            }
+            else
+            {
+                string edgedUri = GetValueFromEnvironment(envVariables, IotEdgedUriVariableName) ?? throw new InvalidOperationException($"Environment variable {IotEdgedUriVariableName} is required.");
+                string gateway = GetValueFromEnvironment(envVariables, GatewayHostnameVariableName);
+
+                Debug.WriteLine("EdgeModuleClientFactory setupTrustBundle from service");
+
                 if (!string.IsNullOrEmpty(gateway))
                 {
-                    IList<X509Certificate2> certs = await _trustBundleProvider.GetTrustBundleAsync(new Uri(edgedUri), DefaultApiVersion).ConfigureAwait(false);
-                    certificateValidator = GetCertificateValidator(certs);
-                    _options.GatewayHostName = gateway;
+                    IList<X509Certificate2> certs = await trustBundleProvider.GetTrustBundleAsync(new Uri(edgedUri), DefaultApiVersion).ConfigureAwait(false);
+                    certificateValidator = GetCertificateValidator(certs, options);
                 }
 
-                return new IotHubModuleClient(CreateInternalClientFromAuthenticationMethod(hostname, authMethod, _options), certificateValidator);
+                return certificateValidator;
             }
         }
 
-        private ICertificateValidator GetCertificateValidator(IList<X509Certificate2> certs)
+        private static ICertificateValidator GetCertificateValidator(IList<X509Certificate2> certs, IotHubClientOptions options)
```

Comment:
```suggestion
        private static ICertificateValidator CreateCertificateValidator(IList<X509Certificate2> certs, IotHubClientOptions options)
```

nit: it is being created, not retrieved

Reply:
static "getters" are a bit dubious, but static "creators" are safe